module-system/sov-state/src/codec.rs

```diff
@@ -0,0 +1,200 @@
+    /// [`Result`] with either the deserialized key or an error.
+    fn try_decode_key(&self, bytes: &[u8]) -> Result<K, Self::KeyError>;
+
+    /// Deserializes a key from a bytes slice.
+    ///
+    /// # Panics
+    /// Panics if the call to [`StateKeyCodec::try_decode_key`] fails. Use
+    /// [`StateKeyCodec::try_decode_key`] if you need to gracefully handle
+    /// errors.
+    fn decode_key(&self, bytes: &[u8]) -> K {
+        self.try_decode_key(bytes)
+            .map_err(|err| {
+                format!(
+                    "Failed to decode key 0x{}, error: {:?}",
+                    hex::encode(bytes),
+                    err
+                )
+            })
+            .unwrap()
+    }
+}
+
+pub trait StateKeyEncodePreservingBorrow<Borrower, Borrowed>
+where
+    Borrowed: ?Sized,
+    Borrower: Borrow<Borrowed>,
+    Self: StateKeyEncode<Borrower> + StateKeyEncode<Borrowed>,
+{
+}
+
+/// A trait for types that can serialize and deserialize values for storage
+/// access.
+pub trait StateValueCodec<V> {
+    /// Error type that can arise during deserialization.
+    type ValueError: std::fmt::Debug;
+
+    /// Serializes a value into a bytes vector.
+    ///
+    /// This method **must** not panic as all instances of the value type are
+    /// supposed to be serializable.
+    fn encode_value(&self, value: &V) -> Vec<u8>;
+
+    /// Tries to deserialize a value from a bytes slice, and returns a
+    /// [`Result`] with either the deserialized value or an error.
+    fn try_decode_value(&self, bytes: &[u8]) -> Result<V, Self::ValueError>;
+
+    /// Deserializes a value from a bytes slice.
+    ///
+    /// # Panics
+    /// Panics if the call to [`StateValueCodec::try_decode_value`] fails. Use
+    /// [`StateValueCodec::try_decode_value`] if you need to gracefully handle
+    /// errors.
+    fn decode_value(&self, bytes: &[u8]) -> V {
+        self.try_decode_value(bytes)
+            .map_err(|err| {
+                format!(
+                    "Failed to decode value 0x{}, error: {:?}",
+                    hex::encode(bytes),
+                    err
+                )
+            })
+            .unwrap()
+    }
+}
+
+/// A market trait for types that implement both [`StateKeyCodec`] and
+/// [`StateValueCodec`].
+///
+/// This trait is blanket-implemented for all types that implement both of the
+/// above.
+pub trait StateCodec<K, V>: StateKeyCodec<K> + StateValueCodec<V> {}
+
+impl<K, V, C> StateCodec<K, V> for C where C: StateKeyCodec<K> + StateValueCodec<V> {}
+
+/// A [`StateCodec`] that uses [`borsh`] for all keys and values.
+#[derive(Debug, Default, PartialEq, Eq, Clone, borsh::BorshDeserialize, borsh::BorshSerialize)]
+pub struct BorshCodec;
+
+impl<K> StateKeyEncode<K> for BorshCodec
+where
+    K: ?Sized + borsh::BorshSerialize,
+{
+    fn encode_key(&self, key: &K) -> Vec<u8> {
+        key.try_to_vec().expect("Failed to serialize key")
```

Comment:
Should we open an issue to make a trait for infallible borsh serialization in the future? While I agree serialization in general should be infallible (except for constrained environments where memory is limited), an implementer of borsh might fail his serialization and we would panic here